### PR TITLE
[Core] Introduce "KeyCredential" base class and AzureKeyCredential extends from it.

### PR DESCRIPTION
**Pull request by mssfang**
Added a base class `KeyCredential` for `AzureKeyCredential` to deviate `Azure` endpoint and credential.

### CODE REVIEW

File: sdk/core/azure-core/CHANGELOG.md [hunk 1]
```diff
@@ -3,6 +3,7 @@
 ## 1.41.0-beta.1 (Unreleased)
 
 ### Features Added
+- Added a base class `KeyCredential` for `AzureKeyCredential` to deviate `Azure` endpoint and credential.
```
Comment (alzimmermsft) [inline sdk/core/azure-core/CHANGELOG.md:6]:
```suggestion
- Added a base class `KeyCredential` for `AzureKeyCredential` for non-`Azure` endpoints and credentials.  
```

**Review by alzimmermsft — APPROVED** [review]

File: sdk/core/azure-core/src/main/java/com/azure/core/credential/AzureKeyCredential.java [hunk 2]
```diff
@@ -23,21 +15,17 @@ public final class AzureKeyCredential {
...
-        this.key = key;
+        super(key);
     }
 
     /**
      * Retrieves the key associated to this credential.
      *
      * @return The key being used to authorize requests.
      */
+    @Override
```
Comment (srnagar) [inline sdk/core/azure-core/src/main/java/com/azure/core/credential/AzureKeyCredential.java:26]: This method can be removed.

File: sdk/core/azure-core/src/main/java/com/azure/core/credential/KeyCredential.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+package com.azure.core.credential;
+
+import com.azure.core.util.logging.ClientLogger;
+
+import java.util.Objects;
+
+/**
+ * Represents a credential that uses a key to authenticate.
+ */
+public class KeyCredential {
```
Comment (srnagar) [inline sdk/core/azure-core/src/main/java/com/azure/core/credential/KeyCredential.java:13]: We should finalize the name with architects. The proposal in arch board was to call this `RotatingKeyCredential`. Although, my preference is to call this `KeyCredential`.

File: sdk/core/azure-core/src/main/java/com/azure/core/credential/KeyCredential.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+     */
+    public String getKey() {
+        return key;
+    }
+
+    /**
+     * Rotates the key associated to this credential.
+     *
+     * @param key The new key to associated with this credential.
+     * @return The updated {@code AzureKeyCredential} object.
```
Comment (srnagar) [inline sdk/core/azure-core/src/main/java/com/azure/core/credential/KeyCredential.java:47]:
```suggestion
     * @return The updated {@code KeyCredential} object.
```

File: sdk/core/azure-core/src/main/java/com/azure/core/http/policy/AzureKeyCredentialPolicy.java [hunk 5]
```diff
@@ -54,48 +44,16 @@ public AzureKeyCredentialPolicy(String name, AzureKeyCredential credential) {
...
-
-    AzureKeyCredentialPolicy(HttpHeaderName name, AzureKeyCredential credential, String prefix) {
-        this.name = name;
-        this.credential = credential;
-        this.prefix = prefix != null ? prefix.trim() : null;
+        super(name, Objects.requireNonNull(credential, "'credential' cannot be null."), prefix);
     }
 
     @Override
     public Mono<HttpResponse> process(HttpPipelineCallContext context, HttpPipelineNextPolicy next) {
```
Comment (srnagar) [inline sdk/core/azure-core/src/main/java/com/azure/core/http/policy/AzureKeyCredentialPolicy.java:51]: These `process*` methods can be deleted.

File: sdk/openai/azure-ai-openai/CHANGELOG.md [hunk 6]
```diff
@@ -5,6 +5,7 @@
 ### Features Added
 
 ### Breaking Changes
+- Removed `NonAzureOpenAIKeyCredential` and `NonAzureOpenAIKeyCredentialPolicy`. Replace by Azure Core `KeyCredential` and `KeyCredentialPolicy`.
```
Comment (srnagar) [inline sdk/openai/azure-ai-openai/CHANGELOG.md:8]: `NonAzureOpenAIKeyCredentialPolicy` was in implementation package. So, that should not be mentioned in changelog.

File: sdk/openai/azure-ai-openai/src/main/java/com/azure/ai/openai/OpenAIClientBuilder.java [hunk 7]
```diff
@@ -189,16 +189,16 @@ public OpenAIClientBuilder credential(AzureKeyCredential azureKeyCredential) {
...
-     * The NonAzureOpenAiKeyCredential used for public OpenAi authentication.
+     * The KeyCredential used for public OpenAi authentication.
      *
-     * @param nonAzureOpenAIKeyCredential The credential for non-azure public OpenAI authenticaton.
+     * @param credential The credential for non-azure public OpenAI authenticaton.
      * @return the object itself.
      */
-    public OpenAIClientBuilder credential(NonAzureOpenAIKeyCredential nonAzureOpenAIKeyCredential) {
-        this.nonAzureOpenAIKeyCredential = nonAzureOpenAIKeyCredential;
+    public OpenAIClientBuilder credential(KeyCredential credential) {
```
Comment (srnagar) [inline sdk/openai/azure-ai-openai/src/main/java/com/azure/ai/openai/OpenAIClientBuilder.java:200]: We should remove `credential(AzureKeyCredential)` overload. We now have one KeyCredential that works for both Azure and OpenAI endpoints. We will need additional changes in this library to determine which endpoint to use and can't rely on `instanceof` check anymore.

**Review by srnagar — CHANGES_REQUESTED** [review]

File: sdk/core/azure-core/src/main/java/com/azure/core/credential/KeyCredential.java (hunk 3 above)
Comment (mssfang) [reply sdk/core/azure-core/src/main/java/com/azure/core/credential/KeyCredential.java:13]: I get @johanste's breath to like the name offline.

File: sdk/core/azure-core/CHANGELOG.md [hunk 8]
```diff
@@ -3,6 +3,7 @@
 ## 1.41.0-beta.1 (Unreleased)
 
 ### Features Added
+- Added a base class `KeyCredential` that `AzureKeyCredential` extends from for `Non-Azure` endpoints and credentials.
```
Comment (srnagar) [inline sdk/core/azure-core/CHANGELOG.md:6]:
```suggestion
- Added a `KeyCredential` to support key-based auth. 
- `AzureKeyCredential` is updated to extend from `KeyCredential`.
```

File: sdk/core/azure-core/src/test/java/com/azure/core/http/policy/AzureKeyCredentialPolicyTests.java [hunk 9]
```diff
@@ -26,10 +26,10 @@ public void setCredential(AzureKeyCredentialPolicy policy, String expectedHeader
     private static Stream<Arguments> setCredentialSupplier() {
         AzureKeyCredential credential = new AzureKeyCredential("asecret");
         return Stream.of(
-            Arguments.of(new AzureKeyCredentialPolicy(HttpHeaderName.AUTHORIZATION, credential, null), "asecret"),
-            Arguments.of(new AzureKeyCredentialPolicy(HttpHeaderName.AUTHORIZATION, credential, "SharedKeyCredential"),
+            Arguments.of(new AzureKeyCredentialPolicy(HttpHeaderName.AUTHORIZATION.toString(), credential, null), "asecret"),
```
Comment (srnagar) [inline sdk/core/azure-core/src/test/java/com/azure/core/http/policy/AzureKeyCredentialPolicyTests.java:29]: We should add separate tests for KeyCredentialPolicy too.

**Review by srnagar — APPROVED** [review]

File: sdk/core/azure-core/src/main/java/com/azure/core/credential/KeyCredential.java (hunk 3 above)
Comment (mssfang) [reply sdk/core/azure-core/src/main/java/com/azure/core/credential/KeyCredential.java:13]: Name was discussed again in the OpenAI archboard and all architects agreed to use this name.

**Pull request merged**